### ISSUE TITLE
SLURM Inconsistency between requested npe_task and number of tasks actually used to run executables

### ISSUE THREAD

**Issue by RussTreadon-NOAA** [NOAA-EMC/global-workflow#1081]
**Expected behavior**
Rocoto jobs should run applications with the `npe_$task` task count specified in `config.resources`

**Current behavior**
Rocoto jobs on Hera (and likely Orion) run applications with the task count equal to `nodes` x `ppn` specified in the experiment xml file

**Machines affected**
Observed on Hera.  Likely occurs on Orion, too

**To Reproduce**
To observe this behavior

1. install g-w on Hera
2. set up EXPDIR
3. populate ROTDIR to start EXPDIR
4. start cycling EXPDIR
5. check job log files for total task count used by apps.   compare this with `npe_$task` in `config.resources`

For example, `config.resources` specifies `npe_anal=84` and `nth_anal=5` for C96 run on Hera.    The Hera `$PSLOT.xml` specifies `<nodes>11:ppn=8:tpp=5</nodes>`   The xml request is correct for Hera.  Each Hera node has 40 slots.  `config.resources` requests 84 tasks with 5 threads per task.   We can fit 8 tasks with 5 threads each per node.  A total of 84 tasks will require 11 nodes with 8 tasks per nodes.   

A check of gdasanal.log shows that `gsi.x` actually ran with 88 tasks.   We find the following printout from `gsi.x` in gdasanal.log
```
87:  INIT_IO:  set IO server task to mype_io=          87
...
44:  GESINFO:  guess jcap_b, namelist jcap_b =        -9999         190
87:  READ_FILES:  analysis date,minutes         2021          12          21
87:           18           0    23127480
87:  READ_FILES:  atm guess file sigf03   3.00000000000000               12
```
Since we start from mpi rank 0, rank 87 is the 88th mpi task.   `gsi.x` is being run with 88 tasks, not the 84 specified in `config.resources`.
   
**Context**
The gdaspostanl failed for a C96 Hera run.   `config.resources` requested 96 tasks with 1 thread per task.  This correctly resulted in `<nodes>3:ppn=40:tpp=1</nodes>` in `$PSLOT.xml`.   Unfortunately, `upp.x` failed with the error message

```
 48:   too many MPI tasks, max is           96  stopping
 44:   too many MPI tasks, max is           96  stopping
 46:   too many MPI tasks, max is           96  stopping
```

Near the top of the `upp.x` runtime output we see that the job actually ran with 120 mpi tasks
```
 95:  ME,NUM_PROCS,NUM_SERVERS=          95         120           0
 43:  ME,NUM_PROCS,NUM_SERVERS=          43         120           0
 97:  ME,NUM_PROCS,NUM_SERVERS=          97         120           0
 99:  ME,NUM_PROCS,NUM_SERVERS=          99         120           0
 59:  ME,NUM_PROCS,NUM_SERVERS=          59         120           0
100:  ME,NUM_PROCS,NUM_SERVERS=         100         120           0
```
120 tasks is consistent with the `$PSLOT.xml` specification of 3 nodes with 40 tasks per node.

**Possible Implementation**
Modify `HERA.env` such that the `APRUN` lines include the number of tasks to use for the given application.  This is already done for some apps.   For example, atmanalrun includes the `-n` option on the APRUN line
```
elif [ $step = "atmanalrun" ]; then

    export CFP_MP=${CFP_MP:-"YES"}
    export USE_CFP=${USE_CFP:-"YES"}
    export APRUNCFP="$launcher -n \$ncmd --multi-prog"

    nth_max=$(($npe_node_max / $npe_node_atmanalrun))

    export NTHREADS_ATMANAL=${nth_atmanalrun:-$nth_max}
    [[ $NTHREADS_ATMANAL -gt $nth_max ]] && export NTHREADS_ATMANAL=$nth_max
    export APRUN_ATMANAL="$launcher -n $npe_atmanalrun"
```

Not sure how easy or hard this is for every `$step` found in `HERA.env`.   Similar changes are likely needed on `ORION.env`.

**Comment by lgannoaa** [comment]
May I know where is the expdir for failed run?

**Comment by WalterKolczynski-NOAA** [comment]
This should've been fixed by #1112

**Comment by KateFriedman-NOAA** [comment]
This issue was indeed resolved by #1112.

A check of a recent cycled experiment on Hera confirms it was resolved. Below I show the correction to the GSI example provided above:

A gdasanal job (with a request of 84 tasks, `npe_anal=84` in `config.resources`) ran with 84 tasks provided to srun (APRUN launcher command):
```
+ exglobal_atmos_analysis.sh[923]: srun -l --export=ALL -n 84 /scratch1/NCEPDEV/stmp2/Kate.Friedman/RUNDIRS/testpmissue/anal.10826/gsi.x
```
Further confirmed later in log (job ran on ranks 0 to 83 = 84 ranks):
```
83:  INIT_IO:  set IO server task to mype_io=          83
```

Closing issue as completed. Thanks again for reporting this issue @RussTreadon-NOAA !
